crud/src/main/java/com/redhat/lightblue/assoc/ResultDoc.java

```diff
@@ -132,64 +139,103 @@ public void setParentDoc(QueryPlanNode parentNode,ChildDocReference ref) {
-                                    refList.add(new ChildDocReference(this,elem.immutableCopy()));
-                                }
-                            } else {
-                                refList.add(new ChildDocReference(this,new Path(cursor.getCurrentKey(),new Path(ref.getName()))));
+                    LOGGER.debug("Iterating {} in {}",p,doc);
+                    KeyValueCursor<Path,JsonNode> cursor=doc.getAllNodes(p);
+                    while(cursor.hasNext()) {
+                        cursor.next();
+                        JsonNode nodeWithRef=cursor.getCurrentValue();
+                        LOGGER.debug("Checking field {}:{}",cursor.getCurrentKey(),nodeWithRef);
+                        if(nodeWithRef instanceof ArrayNode) {
+                            int size=((ArrayNode)nodeWithRef).size();
+                            MutablePath elem=cursor.getCurrentKey().mutableCopy();
+                            int ix=elem.numSegments();
+                            elem.push(0);
+                            elem.push(ref.getName());
+                            for(int i=0;i<size;i++) {
+                                elem.set(ix,i);
+                                refList.add(new ChildDocReference(this,elem.immutableCopy()));
                             }
+                        } else {
+                            refList.add(new ChildDocReference(this,new Path(cursor.getCurrentKey(),new Path(ref.getName()))));
                         }
                     }
-                    children.put(destNode,refList);
                 }
+                children.put(destMd,refList);
             }
         }
         LOGGER.debug("children for {}:{}",id,children);
     }
 
+    /**
+     * This is used to associate the documents when a child document
+     * is searched before the parent document. When this happens, each
+     * reference of the parent document is looked up in child doc list
+     * using the edge criteria.
+     *
+     * @param parentDocs Docs that contain the childDocs
+     * @param childDocs Docs that are contained in parent docs
+     *
+     * childDocs must be all from the same query plan node.
+     * Parentdocs must be all from the same query plan node.
+     * The query plan nodes for child and parent are different, and linked with edgeClauses
+     */
+    public static void associateDocs(List<ResultDoc> parentDocs,
+                                     List<ResultDoc> childDocs,
+                                     List<Conjunct> edgeClauses,
+                                     CompositeMetadata root) {
+        // We have to find a list of childDocs for every reference in a parent doc
+        LOGGER.debug("Associating {} parent docs",parentDocs.size());
+        if(!childDocs.isEmpty()&&!parentDocs.isEmpty()) {
+            QueryPlanNode childDocsNode=childDocs.get(0).node;
+            QueryPlanNode parentDocsNode=parentDocs.get(0).node;
```

Comment:
Is it ever the case with multiple parent or child docs that the query plan node is different for different elements in those lists?  The bindingInfo is created with the plan node from the 0th element.